File: Digital Business Enablement/bluegather/src/main/java/br/com/fiap/bluegather/controller/EventoController.java
package br.com.fiap.bluegather.controller;

import br.com.fiap.bluegather.dto.EventoDTO;
import br.com.fiap.bluegather.dto.EventoResponse;
import br.com.fiap.bluegather.service.EventoService;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import lombok.extern.slf4j.Slf4j;

import jakarta.validation.Valid;

@RestController
@RequestMapping("evento")
@Slf4j
@CrossOrigin(origins = "*", allowedHeaders = "*")
public class EventoController {

    @Autowired
    @Lazy
    private EventoService eventoService;

    @GetMapping
    public List<EventoResponse> listAll() {
        log.info("(" + getClass().getSimpleName() + ") - Buscando todos(as)");
        Set<EventoResponse> eventos = eventoService.listAllEventoResponses();
        return eventos.stream().collect(Collectors.toList());
    }

    @GetMapping("/{id}")
    public EventoResponse findById(@PathVariable Long id) {
        log.info("(" + getClass().getSimpleName() + ") - Exibindo por ID: " + id);
        return eventoService.findEventoResponseById(id);
    }

    @PostMapping
    public ResponseEntity<EventoDTO> create(@RequestBody @Valid EventoDTO newData) {
        log.info("(" + getClass().getSimpleName() + ") - Cadastrando: " + newData);
        return ResponseEntity.status(HttpStatus.CREATED).body(eventoService.create(newData));
    }

    @PutMapping("{id}")
    public ResponseEntity<EventoDTO> update(@PathVariable Long id, @RequestBody @Valid EventoDTO updatedData) {
        log.info("(" + getClass().getSimpleName() + ") - Atualizando por ID: " + id);
        return ResponseEntity.ok(eventoService.update(id, updatedData));
    }

    @DeleteMapping("{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        log.info("(" + getClass().getSimpleName() + ") - Deletando por ID: " + id);
        eventoService.delete(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/titulo/{titulo}")
    public ResponseEntity<Set<EventoDTO>> findByTituloContainingIgnoreCase(@PathVariable String titulo) {
        log.info("(" + getClass().getSimpleName() + ") - Buscando por titulo do evento contendo: " + titulo);
        Set<EventoDTO> list = eventoService.findByTituloContainingIgnoreCase(titulo);
        return ResponseEntity.ok(list);
    }

    @GetMapping("/localizacao/{raio}")
    public ResponseEntity<List<EventoDTO>> findByLocalizacaoERaio(@PathVariable double raio, @RequestBody Map<String, String> coordenadas) {
        double latitude = Double.parseDouble(coordenadas.get("latitude"));
        double longitude = Double.parseDouble(coordenadas.get("longitude"));
        log.info("(" + getClass().getSimpleName() + ") - Buscando eventos dentro de " + raio + " km de " + latitude + ", " + longitude);
        List<EventoDTO> list = eventoService.findByLocalizacaoERaio(latitude, longitude, raio);
        return ResponseEntity.ok(list);
    }

    @GetMapping("/data/{dataInicio}")
    public ResponseEntity<Set<EventoDTO>> findByDataInicioAfter(@PathVariable("dataInicio") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate dataInicio) {
        log.info("(" + getClass().getSimpleName() + ") - Buscando eventos a partir da data de início: " + dataInicio);
        Set<EventoDTO> list = eventoService.findByDataInicioAfter(dataInicio);
        return ResponseEntity.ok(list);
    }

    @GetMapping("/urgencia/{urgencia}")
    public ResponseEntity<Set<EventoDTO>> findByUrgencia(@PathVariable Long urgencia) {
        log.info("(" + getClass().getSimpleName() + ") - Buscando eventos por nível de urgência: " + urgencia);
        Set<EventoDTO> list = eventoService.findByUrgencia(urgencia);
        return ResponseEntity.ok(list);
    }

    @GetMapping("/organizador/{organizadorId}")
    public ResponseEntity<Set<EventoDTO>> findByOrganizadorId(@PathVariable Long organizadorId) {
        log.info("(" + getClass().getSimpleName() + ") - Buscando eventos por organizador: " + organizadorId);
        Set<EventoDTO> list = eventoService.findByOrganizadorId(organizadorId);
        return ResponseEntity.ok(list);
    }

    @GetMapping("/tipo/{tipoEventoId}")
    public ResponseEntity<Set<EventoDTO>> findByTipoEventoId(@PathVariable Long tipoEventoId) {
        log.info("(" + getClass().getSimpleName() + ") - Buscando eventos por tipo de evento: " + tipoEventoId);
        Set<EventoDTO> list = eventoService.findByTipoEventoId(tipoEventoId);
        return ResponseEntity.ok(list);
    }

    @GetMapping("/status/{statusId}")
    public ResponseEntity<Set<EventoDTO>> findByStatusId(@PathVariable Long statusId) {
        log.info("(" + getClass().getSimpleName() + ") - Buscando eventos por status: " + statusId);
        Set<EventoDTO> list = eventoService.findByStatusId(statusId);
        return ResponseEntity.ok(list);
    }

    @GetMapping("/voluntario/{voluntarioId}")
    public ResponseEntity<Set<EventoResponse>> findByVoluntarioId(@PathVariable Long voluntarioId) {
        log.info("(" + getClass().getSimpleName() + ") - Buscando eventos por ID do voluntário: " + voluntarioId);
        Set<EventoResponse> list = eventoService.findEventoResponsesByVoluntarioId(voluntarioId);
        return ResponseEntity.ok(list);
    }
}

File: Enterprise Application Development/bluegather/Controllers/EventoController.cs
using Microsoft.AspNetCore.Mvc;
using BlueGather.Dto;
using BlueGather.Services;

namespace BlueGather.Controllers
{
    [Route("/evento")]
    [ApiController]
    public class EventoController : ControllerBase
    {
        private readonly EventoService _service;

        public EventoController(EventoService service)
        {
            _service = service;
        }

        [HttpGet]
        public async Task<ActionResult<IEnumerable<EventoDto>>> FindAll()
        {
            var list = await _service.FindAll();
            return Ok(list);
        }

        [HttpGet("{id}")]
        public async Task<ActionResult<EventoDto>> FindById(long id)
        {
            try
            {
                var dto = await _service.FindById(id);
                return Ok(dto);
            }
            catch (KeyNotFoundException)
            {
                return NotFound();
            }
        }

        [HttpPost]
        public async Task<ActionResult<EventoDto>> Create([FromBody] EventoDto dto)
        {
            try
            {
                var created = await _service.Create(dto);
                return CreatedAtAction(nameof(FindById), new { id = created.Id }, created);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpPut("{id}")]
        public async Task<ActionResult<EventoDto>> Update(long id, [FromBody] EventoDto updatedEventoDto)
        {
            try
            {
                var updated = await _service.Update(id, updatedEventoDto);
                return Ok(updated);
            }
            catch (KeyNotFoundException)
            {
                return NotFound();
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpDelete("{id}")]
        public async Task<IActionResult> Delete(long id)
        {
            try
            {
                await _service.Delete(id);
                return NoContent();
            }
            catch (KeyNotFoundException)
            {
                return NotFound();
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("urgencia/{urgencia}")]
        public async Task<ActionResult<IEnumerable<EventoDto>>> FindByUrgencia(long urgencia)
        {
            try
            {
                var list = await _service.FindByUrgencia(urgencia);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("organizador/{id}")]
        public async Task<ActionResult<IEnumerable<EventoDto>>> FindByOrganizadorId(long id)
        {
            try
            {
                var list = await _service.FindByOrganizadorId(id);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("tipo/{id}")]
        public async Task<ActionResult<IEnumerable<EventoDto>>> FindByTipoEventoId(long id)
        {
            try
            {
                var list = await _service.FindByTipoEventoId(id);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("status/{id}")]
        public async Task<ActionResult<IEnumerable<EventoDto>>> FindByStatusId(long id)
        {
            try
            {
                var list = await _service.FindByStatusId(id);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("titulo/{titulo}")]
        public async Task<ActionResult<IEnumerable<EventoDto>>> FindByTitulo(string titulo)
        {
            try
            {
                var list = await _service.FindByTitulo(titulo);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("localizacao/{raio}")]
        public async Task<ActionResult<List<EventoDto>>> FindByLocalizacaoERaio(double raio, [FromBody] Dictionary<string, string> coordenadas)
        {
            try
            {
                if (!coordenadas.TryGetValue("latitude", out var latStr) || !double.TryParse(latStr, out var latitude) ||
                    !coordenadas.TryGetValue("longitude", out var lonStr) || !double.TryParse(lonStr, out var longitude))
                {
                    return BadRequest("Coordenadas inválidas.");
                }

                var list = await _service.FindByLocalizacaoERaio(latitude, longitude, raio);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("data/{dataInicio}")]
        public async Task<ActionResult<HashSet<EventoDto>>> FindByDataInicioAfter(DateTime dataInicio)
        {
            try
            {
                var list = await _service.FindByDataInicioAfter(dataInicio);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("voluntario/{id}")]
        public async Task<ActionResult<HashSet<EventoDto>>> FindByVoluntarioId(long id)
        {
            try
            {
                var list = await _service.FindByVoluntarioId(id);
                return Ok(list);
            }
            catch (System.Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }
    }
}